src/protocol/malicious.rs

```diff
@@ -0,0 +1,350 @@
+
+/// A message sent by each helper when they've computed one share of the result
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct UValue<F> {
+    payload: F,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    #[allow(dead_code)]
+    ValidateInput,
+    #[allow(dead_code)]
+    ValidateMultiplySubstep,
+    RevealR,
+    CheckZero,
+}
+
+impl crate::protocol::Step for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::ValidateInput => "validate_input",
+            Self::ValidateMultiplySubstep => "validate_multiply",
+            Self::RevealR => "reveal_r",
+            Self::CheckZero => "check_zero",
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub struct SecurityValidator<'a, F, N> {
+    ctx: ProtocolContext<'a, N>,
+    r_share: Replicated<F>,
+    u: F,
+    w: F,
+}
+
+impl<'a, F: Field, N: Network> SecurityValidator<'a, F, N> {
+    #[allow(dead_code)]
+    pub fn new(ctx: ProtocolContext<'a, N>) -> SecurityValidator<'a, F, N> {
+        let prss = ctx.prss();
+
+        let r_share = prss.generate_replicated(RecordId::from(0));
+        let (u_left, u_right): (F, F) = prss.generate_fields(RecordId::from(1));
+        let (w_left, w_right): (F, F) = prss.generate_fields(RecordId::from(2));
+
+        SecurityValidator {
+            ctx,
+            r_share,
+            u: u_right - u_left,
+            w: w_right - w_left,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn r_share(&self) -> Replicated<F> {
+        self.r_share
+    }
+
+    #[allow(dead_code)]
+    fn compute_dot_product_contribution(a: Replicated<F>, b: Replicated<F>) -> F {
+        (a.left() + a.right()) * (b.left() + b.right()) - a.right() * b.right()
+    }
+
+    #[allow(dead_code)]
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn accumulate_macs(
+        &mut self,
+        ctx: ProtocolContext<'_, N>,
+        record_id: RecordId,
+        x: Replicated<F>,
+        rx: Replicated<F>,
+    ) {
+        // The helpers need to use the same shared randomness to generate the random constant used to validate a given multiplication.
+        // This is a bit tricky, because we cannot count on the multiplications being executed in the same order across all the helpers.
+        // The easiest way is to just narrow the context used to perform the multiplication, and then re-use the same record_id.
+        // This ensures that when the helpers all go to validate the multiplication: "1/foo/bar/baz", they all use the prss from "1/foo/bar/baz/validate".
+        // That way, we don't need to worry about the order in which the multiplications are executed.
+        let random_constant = ctx.prss().generate_replicated(record_id);
+
+        self.u += Self::compute_dot_product_contribution(random_constant, rx);
+        self.w += Self::compute_dot_product_contribution(random_constant, x);
+    }
+
+    #[allow(dead_code)]
```

Comment:
Given how many of these you have, it might be easier to put `#![allow(dead_code)]` at the top of the module.